src/notebooks/flatten_and_explode.ipynb

```diff
@@ -0,0 +1,524 @@
+       "datasetInfos": [],
+       "metadata": {},
+       "removedWidgets": [],
+       "type": "html"
+      }
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "'''\n",
+    "A wrapper function to perform full flattening of an input spark data frame.\n",
+    "Struct columns are separated out into multiple columns, and then the intermediate\n",
+    "expanded data frame is exploded so that each new element encountered generates a \n",
+    "new row.\n",
+    "'''\n",
+    "def flatten_df(dfflat):\n",
+    "    while len([field.name for field in dfflat.schema.fields if type(field.dataType) == StructType or type(field.dataType) == ArrayType ]) !=0 :\n",
+    "        dfflat = flatten_structs(dfflat)\n",
+    "        dfflat = explode_arrays(dfflat)\n",
+    "    return dfflat"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 0,
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "inputWidgets": {},
+     "nuid": "1af9907f-162a-45d6-98e9-7f7025571bd2",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/html": [
+       "<style scoped>\n",
+       "  .ansiout {\n",
+       "    display: block;\n",
+       "    unicode-bidi: embed;\n",
+       "    white-space: pre-wrap;\n",
+       "    word-wrap: break-word;\n",
+       "    word-break: break-all;\n",
+       "    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n",
+       "    font-size: 13px;\n",
+       "    color: #555;\n",
+       "    margin-left: 4px;\n",
+       "    line-height: 19px;\n",
+       "  }\n",
+       "</style>\n",
+       "<div class=\"ansiout\"></div>"
+      ]
+     },
+     "metadata": {
+      "application/vnd.databricks.v1+output": {
+       "addedWidgets": {},
+       "arguments": {},
+       "data": "<div class=\"ansiout\"></div>",
+       "datasetInfos": [],
+       "metadata": {},
+       "removedWidgets": [],
+       "type": "html"
+      }
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "'''\n",
+    "Generates a DDL \"create-table\" string from a flattened spark dataframe. The result\n",
+    "of this function is a statement of the form:\n",
+    "\n",
+    "  CREATE TABLE [table_name] (\n",
+    "    [field1] nvarchar(50),\n",
+    "    [field2] nvarchar(50),\n",
+    "    ...\n",
+    "  )\n",
+    "  \n",
+    "For type safety reasons, all fields in the created DDL are saved as nvarchar(50).\n",
+    "'''\n",
+    "def generate_ddl(df, tblname):\n",
+    "    createtbl = 'CREATE TABLE [' + tblname +  '] ( \\n'\n",
+    "    num_columns = len(dfflat.columns)\n",
+    "    for i, y in enumerate(dfflat.columns):\n",
+    "        if i == num_columns-1:\n",
+    "            column_name = '\\t['+ y + '] nvarchar(50)); \\n\\n'\n",
+    "        else:\n",
+    "            column_name = '\\t[' + y + '] nvarchar(50), \\n'\n",
+    "        createtbl += column_name\n",
```

Comment:
I would replace lines 340 to 346 with the following:

```
fields = ", \n".join([f"\t[{column}] nvharchar(50)" for column in dfflat.columns])
createtbl = "\n".join([createtbl, fields, ";"])
```

This approach appends the semi-colon on a newline by itself, but SQL doesn't care about that and it improves the readability of the code.